### PR TITLE
[IMP] stock: record expiry date for incoming lot

### PR DESCRIPTION
**Pull request by dut-odoo**
Description of the issue/feature this PR addresses:
Task: https://www.odoo.com/web#id=28937&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720
Pad: https://pad.odoo.com/p/r.5744fd92c97ec0f29efc9c9b7ad64a9f

Current behavior before PR:

Desired behavior after PR is merged:
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/product_expiry/models/production_lot.py [hunk 1]
```diff
@@ -34,21 +34,38 @@ def _get_dates(self, product_id=None):
...
         for field, value in dates_dict.items():
             setattr(self, field, value)
+
+
+class PackOperationLot(models.Model):
+    _inherit = "stock.pack.operation.lot"
+
+    def _default_life_date(self):
+        if self.env.context.get('product_id'):
+            life_days = self.env['product.product'].browse(self.env.context['product_id']).life_time
```
Comment (tivisse) [inline addons/product_expiry/models/production_lot.py]: We should place a product_id in the context. You should find another way to achieve this, like default_product_id or something like this.

File: addons/product_expiry/views/production_lot_views.xml [hunk 2]
```diff
@@ -35,4 +35,14 @@
         </field>
      </record>
 
+    <record id="stock_pack_operation_view_form_inherit_product_expiry" model="ir.ui.view">
+        <field name="name">stock.pack.operation.form.inherit.product.expiry</field>
+        <field name="model">stock.pack.operation</field>
+        <field name="inherit_id" ref="stock.view_pack_operation_lot_form" />
+        <field name="arch" type="xml">
+            <field name="lot_name" position="after">
+                <field name="life_date" invisible="not context.get('only_create', False)"/>
```
Comment (tivisse) [inline addons/product_expiry/views/production_lot_views.xml]: No.

File: addons/product_expiry/models/production_lot.py [hunk 3]
```diff
@@ -23,7 +23,7 @@ def _compute_product_expiry_alert(self):
         for lot in self.filtered(lambda l: l.alert_date):
             lot.product_expiry_alert = lot.alert_date <= current_date
 
-    def _get_dates(self, product_id=None):
+    def _get_dates(self, product_id=None, life_date=None):
```
Comment (tivisse) [inline addons/product_expiry/models/production_lot.py]: Is it really necessary to modify the method signature?

File: addons/product_expiry/models/production_lot.py [hunk 4]
```diff
@@ -34,21 +34,38 @@ def _get_dates(self, product_id=None):
...
-                    date = datetime.datetime.now() + datetime.timedelta(days=duration)
+                    date = op_date + datetime.timedelta(days=duration)
                     res[field] = fields.Datetime.to_string(date)
         return res
 
     # Assign dates according to products data
     @api.model
     def create(self, vals):
-        vals.update(self._get_dates(vals.get('product_id')))
+        vals.update(self._get_dates(vals.get('product_id'), vals.get('life_date')))
```
Comment (mdi-odoo) [inline addons/product_expiry/models/production_lot.py]: The method get() returns a value for the given key. If key is not available then returns default value None.

File: addons/product_expiry/models/production_lot.py (hunk 1 above)
Comment (mdi-odoo) [inline addons/product_expiry/models/production_lot.py]: Already placed a product_id in the context: https://github.com/odoo/odoo/pull/13852/files#diff-c1db5239e44b020e0202bb0257f8a1f3R83 .

File: addons/product_expiry/models/production_lot.py (hunk 3 above)
Comment (dut-odoo) [inline addons/product_expiry/models/production_lot.py]: Yes, need to modify the method signature

File: addons/product_expiry/models/production_lot.py (hunk 1 above)
Comment (mdi-odoo) [inline addons/product_expiry/models/production_lot.py]: Actually **product_id** field is not present in current model, so we need to pass it from context.

File: addons/product_expiry/views/production_lot_views.xml (hunk 2 above)
Comment (mdi-odoo) [inline addons/product_expiry/views/production_lot_views.xml]: Why ?  Please suggest if you have any good solution.

File: addons/product_expiry/models/production_lot.py (hunk 3 above)
Comment (bst-odoo) [inline addons/product_expiry/models/production_lot.py]: I think @tivisse wanted an explanation why it is necessary to do it ;)

File: addons/product_expiry/views/production_lot_views.xml (hunk 2 above)
Comment (bst-odoo) [inline addons/product_expiry/views/production_lot_views.xml]: @tivisse what should be done instead ?

File: addons/stock/models/stock_picking.py [hunk 5]
```diff
@@ -900,7 +900,10 @@ def _create_lots_for_picking(self):
         Lot = self.env['stock.production.lot']
         for pack_op_lot in self.mapped('pack_operation_ids').mapped('pack_lot_ids'):
             if not pack_op_lot.lot_id:
-                lot = Lot.create({'name': pack_op_lot.lot_name, 'product_id': pack_op_lot.operation_id.product_id.id})
+                lot_values = {'name': pack_op_lot.lot_name, 'product_id': pack_op_lot.operation_id.product_id.id}
```
Comment (hmo-odoo) [inline addons/stock/models/stock_picking.py]: life_date field is introduced from product_expiry module.

Please do not write stuff regarding life_date field stuff

File: addons/product_expiry/tests/test_lot_dates.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+from datetime import timedelta
+
+from odoo import fields
+from odoo.tests.common import TransactionCase
+
+
+class TestLotDates(TransactionCase):
+
+    def setUp(self):
+        super(TestLotDates, self).setUp()
```
Comment (mdi-odoo) [inline addons/product_expiry/tests/test_lot_dates.py]: No need of this. Please remove it.

File: addons/product_expiry/tests/test_lot_dates.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    def test_00_lot_dates(self):
+
+        # locations
+        location_id = self.ref('stock.stock_location_suppliers')
+        location_dest_id = self.ref('stock.stock_location_stock')
+        today = fields.datetime.today()
```
Comment (mdi-odoo) [inline addons/product_expiry/tests/test_lot_dates.py]: Use **fields.Datetime.now()** instead .

**Comment by pimodoo** [comment]
Sorry this task has been cancelled: https://www.odoo.com/web#id=28937&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720

**Pull request closed without merging**